File: src/dataStructure/stack/StackDemo.java
package dataStructure.stack;

public class StackDemo {

    private static int top = -1;
    private static int[] ints = new int[10];

    public static void main(String[] arg) {

        for (int i = 0; i < 10 ; i++) {
            push(i);
        }

        for (int i = 1; top >= 0; i++) {
            System.out.println("Data pos:"+i+" value:"+ pop());
        }

        //now stack is empty
        //add new data
        push(5);
        push(4);

        System.out.println("Get new data:"+ pop());
    }

    private static void push(int n) {
        top += 1;
        ints[top] = n;
    }

    private static int pop(){

        if (top < 0){
            return -1;
        }

        int ret = ints[top];
        top -= 1;
        return ret;
    }
}

File: src/dataStructure/stack/StackDemo.kt
package dataStructure.stack

fun main(args: Array<String>) {

    val stack = Stack<Int>()

    for (i in 0 until 10){
        stack.push(i)
    }

    println("Stack length:${stack.length}")

    while (stack.length > 0){
        println("Stack: value:${stack.pop()} length:${stack.length}")
    }

    //get data on empty stack
    //println("Stack ${stack.pop()}")
}

class Stack<T> {
    private var top = -1
    private val array:MutableList<T> = mutableListOf()

    fun push(value: T) {
        top += 1
        array.add(top,value)
    }

    fun pop(): T {

        if (top == -1) {
            throw Exception("Stack is empty")
        }

        val value = array[top]
        top -= 1
        return value
    }

    val length:Int get() {
        return if (top == -1) 0
        else top + 1
    }

}